packages/flame_studio/lib/src/core/component_tree.dart

```diff
@@ -0,0 +1,79 @@
+import 'package:flame/components.dart';
+import 'package:flame/game.dart';
+import 'package:flame_studio/src/core/game_controller.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+final componentTreeProvider =
+    StateNotifierProvider<ComponentTreeObserver, ComponentTreeState>((ref) {
+  final gameState = ref.watch(gameControllerProvider);
+  return ComponentTreeObserver(gameState.game as FlameGame?);
+});
+
+@immutable
+class ComponentTreeNode {
+  const ComponentTreeNode._(this.component, this.children);
+
+  factory ComponentTreeNode.fromComponent(Component component) {
+    return ComponentTreeNode._(
+      component,
+      component.hasChildren
+          ? component.children.map(ComponentTreeNode.fromComponent).toList()
+          : null,
```

Comment:
any reason to make this distinction between null and empty list?

Reply:
If a component doesn't have children, then I don't want to instantiate its children ComponentSet unnecessarily.

Reply:
The majority of components don't have any children, so why create empty lists unnecessarily?